pkg/quota/v1/evaluator/core/persistent_volume_claims.go

```diff
@@ -235,3 +272,59 @@ func RequiresQuotaReplenish(pvc, oldPVC *corev1.PersistentVolumeClaim) bool {
 	}
 	return false
 }
+
+// pvcMatchesScopeFunc is a function that knows how to evaluate if a pvc matches a scope
+func pvcMatchesScopeFunc(selector corev1.ScopedResourceSelectorRequirement, object runtime.Object) (bool, error) {
+	pvc, err := toExternalPersistentVolumeClaimOrError(object)
+	if err != nil {
+		return false, err
+	}
+
+	switch selector.ScopeName {
+	case corev1.ResourceQuotaScopeVolumeAttributesClass:
+		if selector.Operator == corev1.ScopeSelectorOpExists {
+			// This is just checking for existence of a volumeAttributesClass on the pvc,
+			// no need to take the overhead of selector parsing/evaluation.
+			if ptr.Deref(pvc.Spec.VolumeAttributesClassName, "") != "" {
+				return true, nil
+			}
+			if ptr.Deref(pvc.Status.CurrentVolumeAttributesClassName, "") != "" {
+				return true, nil
+			}
+			modifyStatus := pvc.Status.ModifyVolumeStatus
+			if modifyStatus != nil && modifyStatus.TargetVolumeAttributesClassName != "" {
+				return true, nil
+			}
+			return false, nil
+		}
+		return pvcMatchesSelector(pvc, selector)
+	}
+	return false, nil
+}
+
+func pvcMatchesSelector(pvc *corev1.PersistentVolumeClaim, selector corev1.ScopedResourceSelectorRequirement) (bool, error) {
+	labelSelector, err := helper.ScopedResourceSelectorRequirementsAsSelector(selector)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse and convert selector: %v", err)
+	}
+
+	vacNames := sets.New[string]()
+	if ptr.Deref(pvc.Spec.VolumeAttributesClassName, "") != "" {
+		vacNames.Insert(*pvc.Spec.VolumeAttributesClassName)
+	}
+	if ptr.Deref(pvc.Status.CurrentVolumeAttributesClassName, "") != "" {
+		vacNames.Insert(*pvc.Status.CurrentVolumeAttributesClassName)
+	}
+	modifyStatus := pvc.Status.ModifyVolumeStatus
+	if modifyStatus != nil && modifyStatus.TargetVolumeAttributesClassName != "" {
+		vacNames.Insert(modifyStatus.TargetVolumeAttributesClassName)
+	}
```

Comment:
we quota spec, but we avoid quota-ing status. The actual state should always be allowed to be written.

Reply:
Even if we only quota spec field, we will still have this trouble https://github.com/kubernetes/kubernetes/issues/124436. 

For example, 

Firstly, we create 2 quota, one for gold and one for silver. 
Secondly, we create a pvc with volumeAttributesClass=gold, the gold quota will be used.
Thirdly, we update the pvc with volumeAttributesClass=silver, the expected behavior is that the silver quota will be used and the gold quota will be released. In fact, no change will happen.

@deads2k

Reply:
/cc @xing-yang @sunnylovestiramisu @msau42

Reply:
If we don't quota status field, users cab create unlimited volumes with special attributes by changing a PVC's vac into a non-existent vac. Those modified PVCs have the same attributes as the original PVCs. 

Should we introduce a new resource name for this feature like storageclass instead of using the scope mechanism?

-  new resource name is implemented in https://github.com/kubernetes/kubernetes/pull/121895